File: solutions/java/src/digitalwallet/CurrencyConverter.java
package digitalwallet;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.HashMap;
import java.util.Map;

public class CurrencyConverter {
    private static final Map<Currency, BigDecimal> exchangeRates = new HashMap<>();

    static {
        // Initialize exchange rates
        exchangeRates.put(Currency.USD, BigDecimal.ONE);
        exchangeRates.put(Currency.EUR, new BigDecimal("0.85"));
        exchangeRates.put(Currency.GBP, new BigDecimal("0.72"));
        exchangeRates.put(Currency.JPY, new BigDecimal("110.00"));
        // Add more exchange rates as needed
    }

    public static BigDecimal convert(BigDecimal amount, Currency sourceCurrency, Currency targetCurrency) {
        BigDecimal sourceRate = exchangeRates.get(sourceCurrency);
        BigDecimal targetRate = exchangeRates.get(targetCurrency);
        return amount.multiply(sourceRate).divide(targetRate, RoundingMode.HALF_UP);
    }
}

File: solutions/golang/digitalwalletservice/currency_converter.go
package digitalwallet

import (
	"math/big"
	"sync"
)

type CurrencyConverter struct {
	exchangeRates map[Currency]*big.Float
	mu            sync.RWMutex
}

var (
	currencyConverter *CurrencyConverter
	converterOnce     sync.Once
)

func GetCurrencyConverter() *CurrencyConverter {
	converterOnce.Do(func() {
		currencyConverter = &CurrencyConverter{
			exchangeRates: make(map[Currency]*big.Float),
		}
		currencyConverter.initializeRates()
	})
	return currencyConverter
}

func (cc *CurrencyConverter) initializeRates() {
	cc.mu.Lock()
	defer cc.mu.Unlock()

	one := big.NewFloat(1)
	cc.exchangeRates[USD] = one
	cc.exchangeRates[EUR] = big.NewFloat(0.85)
	cc.exchangeRates[GBP] = big.NewFloat(0.72)
	cc.exchangeRates[JPY] = big.NewFloat(110.00)
}

func (cc *CurrencyConverter) Convert(amount *big.Float, sourceCurrency, targetCurrency Currency) *big.Float {
	cc.mu.RLock()
	defer cc.mu.RUnlock()

	sourceRate := cc.exchangeRates[sourceCurrency]
	targetRate := cc.exchangeRates[targetCurrency]

	result := new(big.Float).Mul(amount, sourceRate)
	return new(big.Float).Quo(result, targetRate)
}
